### DISCUSSION TITLE
CBOR with Berry?

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#18537] [Q&A]
Hello, wondering if [CBOR](https://datatracker.ietf.org/doc/html/rfc8949) can be imported somehow to be used with Berry Script? I want to send data using JSON payloads, but store in the SD card in case of failure using CBOR for space saving (see picture for example that was generated from [here](https://cbor.me/))

If not, would it be hard to create a function accessible inside Berry that takes a JSON object, and returns a CBOR string pointer that can be then saved or whatever?

![image](https://user-images.githubusercontent.com/6442613/235300704-8c4da093-e578-4bd7-9fe0-6c76c9b226f7.png)

**Comment by barbudor** [comment]
Interresting
I knew about protobuff as a similar approach but this one seems more optimized for low systems

More than imported it's a matter of implementation
If there is an existing implementation in Berry for sure it could be imported
As well a C library could be encapsulated in a Berry API

Any hints?

**Comment by sfromis** [comment]
With Tasmota+Berry you have good features for subscribing to arbitrary MQTT payloads, and do JSON parsing when relevant, or code against a binary format. The CBOR format is not included, but would probably be feasible to code in the Berry language, if you think that this is useful.

If what you show is supposed to be a compression of what you have in the JSON data, it does really not look like much compression. Much worse than a simple gz.

**Comment by s-hadinger** [comment]
The space savings do not look spectacular in the example above : -22% while requiring processing and memory. 

You could try unishox, while its not specifically meant for json, it could give decent results. But again I'm not sure it's worth the memory usage while SD storage is multiple orders of magnitude bigger than memory

**Reply by sfromis** [reply]
Well, I found the unishox module, but was reluctant to recommend it:
```
import unishox
s = '{"Time":"2023-04-29T14:37:03.999+02:00","ESP32":{"HallEffect":34}}'
print(size(s), size(unishox.compress(s)))

BRY: Exception> 'internal_error' - unishox size=62 ret=-1
stack traceback:
	<native>: in native function
	<unknown source>: in function `main`
```
It did not help repeating the data to make sure that it was compressible.

**Reply by s-hadinger** [reply]
Hmmm. Not good

**Reply by s-hadinger** [reply]
You can try with online with https://unishox.hadinger.fr/

**Reply by sfromis** [reply]
The online page had no trouble with the same string, but without much in repeats, it only managed to squeeze out 2 bytes. Unsurprisingly, much better result with the string doubled.

**Reply by s-hadinger** [reply]
Unishox shines for lowercase text, numbers and repetitions. All other characters are more costly

**Reply by sfromis** [reply]
Assuming that Unishox comes to work in Berry as in the web page, it should be able to cut the sample JSON text to about 2/3 size.

**Reply by Slyke** [reply]
It could be useful for when bandwidth is tight. JSON is great, but if each byte counts something that compresses is useful. Not exactly my use case, but could see how it would benefit others. Unishox would be great as well if CBOR is not the correct flavor for Tasmota.

**Reply by s-hadinger** [reply]
In such case you need support on both ends which narrows down even further the use case. I'm still not convinced that the additional processing and memory is worth the small compression factor

**Reply by sfromis** [reply]
As Donald Knuth said, "premature optimization is the root of all evil". Doing extra work to create complexity for obtaining small and mostly unimportant savings is better skipped unless having a case where it demonstrably really matters. Then the cost of complexity may be worth it, but not until then. In the Tasmota context, running on ESP8266 has been a case where byte counting matters.

I remember a time in the past, where every byte actually counted (as opposed to now). When I protested that a design would have a problem in year 2000, due to it storing year in two digits, the response was that it did not matter, as the code would be long replaced by then. It was not :grin:

**Reply by Jason2866** [reply]
> It could be useful for when bandwidth is tight. JSON is great, but if each byte counts something that compresses is useful. Not exactly my use case, but could see how it would benefit others. Unishox would be great as well if CBOR is not the correct flavor for Tasmota.

Such a slow bandwidth, where this would be an problem, is not supported from Tasmota. There will be no code implemented in Tasmota for seldom corner cases. Might be considered to change if a real life use case can be shown.
